### PR TITLE
Extend font locking.

### PR DESCRIPTION
**Pull request by tytus-metrycki**
- Rewrite of font locking. 
   I've rewritten regexs' into separate lists. They should be easier to extend in the future.
   I also added some keywords (string manipulation and aggregation) and removed some (such as ":-" "[" "]") 
- Inherit from prog-mode instead of fundamental-mode
- Autoload for files with extensions ".dl"

### CODE REVIEW

**Comment by karljs** [comment]
I'm not affiliated with the Souffle project at all, but this looks great and I hope it gets merged.  It looks like you're missing functor, and it seems to highlight `symbol` and `number` at a higher priority `.symbol_type` and `.number_type`, which results in something like:

.**symbol**_type

I don't actually think this is a regression, though, so I'd encourage folks to not consider it a blocker.

Thanks for working on it!

**Comment by tytus-metrycki** [comment]
@karljs, thank you for the kind words.

I think highlight works correctly now - it didn't earlier because there was no highlighting of `.*_type`. I think emacs tries the rules in the order they are defined, so it should be fine. Screenshot:
Screenshot 2019-11-23 at 12 19 41

Adding a functor is a good idea :)
If this pull request gets accepted I will add it as well as other missing keywords (bit/logical functions).

File: souffle-mode.el [hunk 1]
```diff
@@ -13,25 +13,83 @@
...
+    (list "min" "max" "sum" "count")
+    "Souffle builtin aggregation functions.")
+
+(defconst souffle-types
+    (list "symbol" "number")
+    "Souffle builtin types.")
+
+(defconst souffle-highlights
+    (let* (
+              ;; Generate regexp fox each category.
```
Comment (langston-barrett) [inline souffle-mode.el]:
```suggestion
              ;; Generate regexp for each category.
```

File: souffle-mode.el [hunk 2]
```diff
@@ -13,25 +13,83 @@
...
+(defconst souffle-string-functions
+    (list "cat" "contains" "match" "ord" "strlen" "substr" "to_number" "to_string")
+    "Souffle builtin string functions.")
+
+(defconst souffle-aggregation-functions
+    (list "min" "max" "sum" "count")
+    "Souffle builtin aggregation functions.")
+
+(defconst souffle-types
+    (list "symbol" "number")
```
Comment (langston-barrett) [inline souffle-mode.el]: https://souffle-lang.github.io/types
```suggestion
    (list "symbol" "number" "unsigned" "float")
```

File: souffle-mode.el [hunk 3]
```diff
@@ -13,25 +13,83 @@
...
+             (,souffle-types-regexp . font-lock-type-face)
+             )
+        ))
+
+
+
+;;;;;;;;;;;;;;;;;;;;;;;
+;; define major mode ;;
+;;;;;;;;;;;;;;;;;;;;;;;
+(define-derived-mode souffle-mode prog-mode "souffle"
```
Comment (langston-barrett) [inline souffle-mode.el:80]: Ah, we had the same thought :smile: #3

**Review by langston-barrett — APPROVED** [review]
Agreed with @karljs, this seems like a great addition. @b-scholz Any chance we could get this merged?

**Comment by b-scholz** [comment]
I overlooked this one. Can somebody resolve the conflict? I am happy to approve it.

**Comment by tytus-metrycki** [comment]
@b-scholz done!

**Pull request merged**